src/main/java/net/minecraftforge/common/util/SortedRegistry.java

```diff
@@ -0,0 +1,219 @@
+        var lastIndex = last != null ? elements.indexOf(last) : elements.size();
+        if (firstIndex >= lastIndex) return ImmutableList.of();
+        return elements.subList(firstIndex + 1, lastIndex);
+    }
+
+    public static class Builder<T>
+    {
+        private final Class<T> type;
+        private final BiMap<ResourceLocation, T> entries = HashBiMap.create();
+        private final Multimap<ResourceLocation, ResourceLocation> edges = HashMultimap.create();
+
+        public Builder(Class<T> type)
+        {
+            this.type = type;
+        }
+
+        /**
+         * Adds an element to the sorted registry.
+         * @param element The element
+         * @param name The name of the element
+         * @param predecessors A list of elements that must appear before this element (String, ResourceName, and T are allowed)
+         * @param successors A list of elements that must appear after this element (String, ResourceName, and T are allowed)
+         * @param <T2> The type of the element
+         * @return The element
+         */
+        public synchronized <T2 extends T> T2 add(T2 element, ResourceLocation name, List<Object> predecessors, List<Object> successors)
+        {
+            Objects.requireNonNull(predecessors, "Predecessor list may not be null.");
+            Objects.requireNonNull(successors, "Successor list may not be null.");
+            if (entries.containsKey(name))
+                throw new IllegalArgumentException("An entry has already been registered with this name: " + name);
+            entries.put(name, element);
+            for(var other : predecessors)
+                edges.put(getName(other), name);
+            for(var other : successors)
+                edges.put(name, getName(other));
+            return element;
+        }
+
+        private ResourceLocation getName(Object element)
+        {
+            Objects.requireNonNull(element, "Attempted to get name of a null object.");
+            if (element instanceof ResourceLocation rl)
+                return rl;
+            if (element instanceof String s)
+                return new ResourceLocation(s);
+            if (type.isAssignableFrom(element.getClass()))
+            {
+                var name = entries.inverse().get(type.cast(element));
+                return Objects.requireNonNull(name, "Cannot add a dependency on an element that is not registered.");
+            }
+            throw new IllegalStateException("Invalid object type passed into dependencies: " + element.getClass());
+        }
+
+        /**
+         * Creates a new {@link SortedRegistry} from the registered elements and dependencies.
+         * @return A new registry
+         */
+        @SuppressWarnings("UnstableApiUsage")
+        public SortedRegistry<T> build()
+        {
+            var graph = GraphBuilder.directed().nodeOrder(ElementOrder.<T>insertion()).build();
+            for(var element : entries.values())
```

Comment:
Formatting: space before the opening parentheses in a for-expression:
```suggestion
            for (var element : entries.values())
```